posts/2023-10-11-the-quest-for-a-smooth-onboarding.adoc

```diff
@@ -0,0 +1,138 @@
+
+== Why do we want an easier onboarding?
+
+There is no real debate about whether we actually want developers to get started faster. The benefit is obvious: the sooner a developer can contribute, the sooner they produce value.
+
+At the same time, once a team knows the ins and outs of a system, they don't see the quirks and implicit conventions that hinder onboarding. It does not even slow them down, because they installed all the tooling and included everything in their daily routine. The benefit of an easy onboarding becomes less obvious. Why invest time on this, while we have features to do and the team is going full speed?
+
+Still, there is value in a good onboarding, for a few reasons:
+
+*We won't always be there as a contractor.* The reality of being a contractor is that one day the project ends, no matter how well it went. Maybe there is nothing to add to the application, maybe the customer ran out of funds for this year. In both cases, the goal is that the customer can be autonomous and take over if needed.
+
+*The application might not need a full-time team.* A lot of systems don't need new features every week. Maybe the customer will ask for a new webpage, a bugfix, or an upgrade from time to time, but overall the project is in "maintenance mode". No one will be fully focused on that particular application most of the time. And a developer will have to do some kind of new onboarding every time to refresh their memory.
+A developer might go away. At an individual level, life happens. Someone can get a better opportunity and leave the company, or they need to take a long sick leave, or they get hit by the https://en.wikipedia.org/wiki/Bus_factor[proverbial bus]. Then another person has to come in and fill the gap.
+
+*A smooth onboarding means a happy productive developer.* This one is more arguable as it's harder to measure. It's similar to learning about a new product, say a shiny new database. If the "getting started" page's instructions work immediately then it inspires confidence. On the other hand, if the first command fails, and you have to search on Stackoverflow to get it working, it's irritating and does not convey a good image. The gist of it is that if you spend your day chasing information about how to start the app, by the time you get ready to actually code, you have exhausted your focus.
+
+== Documentation is not a silver bullet
+
+The first thing that comes to mind at this point is that we should write documentation. At the risk of sounding a bit provocative, documentation is a necessary evil. It should be our last resort, yet we often start with documentation. It is only one component of the equation, and there is often too much emphasis on it.
```

Comment:
`It should be our last resort` I really don't agree with this sentence.
As you mentioned in the beginning the documentation should cover the code, the context and the domain.
I believe the context and a bit the domain have to be covered with documentation. The basic setup to be able to development the project as well (main stack and tool that have to be installed). Interaction between very separate components, for example, if appliable.
But for the rest, mostly the code, you can argue that making an effort of automating the documentation from the code is a valuable effort in the middle and long term because is something that is hard to keep up-to-date and it depends on the way the codebase evolves.